src/linux.rs

```diff
@@ -0,0 +1,380 @@
+    c_int, AF_INET, AF_INET6, AF_NETLINK, AF_UNSPEC, ARPHRD_NONE, IFLA_IFNAME, IFLA_MTU,
+    NETLINK_ROUTE, NLMSG_ERROR, NLM_F_ACK, NLM_F_REQUEST, RTA_DST, RTA_OIF, RTM_GETLINK,
+    RTM_GETROUTE, RTM_NEWLINK, RTM_NEWROUTE, RTN_UNICAST, RT_SCOPE_UNIVERSE, RT_TABLE_MAIN,
+};
+use static_assertions::{const_assert, const_assert_eq};
+
+use crate::{aligned_by, default_err, routesocket::RouteSocket, unlikely_err};
+
+#[allow(
+    clippy::struct_field_names,
+    non_camel_case_types,
+    clippy::too_many_lines
+)]
+mod bindings {
+    include!(env!("BINDINGS"));
+}
+
+#[allow(clippy::cast_possible_truncation)] // Guarded by the following `const_assert_eq!`.
+const AF_INET_U8: u8 = AF_INET as u8;
+const_assert_eq!(AF_INET_U8 as i32, AF_INET);
+
+#[allow(clippy::cast_possible_truncation)] // Guarded by the following `const_assert_eq!`.
+const AF_INET6_U8: u8 = AF_INET6 as u8;
+const_assert_eq!(AF_INET6_U8 as i32, AF_INET6);
+
+#[allow(clippy::cast_possible_truncation)] // Guarded by the following `const_assert_eq!`.
+const AF_UNSPEC_U8: u8 = AF_UNSPEC as u8;
+const_assert_eq!(AF_UNSPEC_U8 as i32, AF_UNSPEC);
+
+#[allow(clippy::cast_possible_truncation)] // Guarded by the following `const_assert_eq!`.
+const NLM_F_REQUEST_U16: u16 = NLM_F_REQUEST as u16;
+const_assert_eq!(NLM_F_REQUEST_U16 as c_int, NLM_F_REQUEST);
+
+#[allow(clippy::cast_possible_truncation)] // Guarded by the following `const_assert_eq!`.
+const NLM_F_ACK_U16: u16 = NLM_F_ACK as u16;
+const_assert_eq!(NLM_F_ACK_U16 as c_int, NLM_F_ACK);
+
+#[allow(clippy::cast_possible_truncation)] // Guarded by the following `const_assert_eq!`.
+const NLMSG_ERROR_U16: u16 = NLMSG_ERROR as u16;
+const_assert_eq!(NLMSG_ERROR_U16 as c_int, NLMSG_ERROR);
+
+const_assert!(size_of::<nlmsghdr>() <= u8::MAX as usize);
+const_assert!(size_of::<rtmsg>() <= u8::MAX as usize);
+const_assert!(size_of::<rtattr>() <= u8::MAX as usize);
+const_assert!(size_of::<ifinfomsg>() <= u8::MAX as usize);
+
+const NETLINK_BUFFER_SIZE: usize = 8192; // See netlink(7) man page.
+
+#[repr(C)]
+enum AddrBytes {
+    V4([u8; 4]),
+    V6([u8; 16]),
+}
+
+impl AddrBytes {
+    const fn new(ip: IpAddr) -> Self {
+        match ip {
+            IpAddr::V4(ip) => Self::V4(ip.octets()),
+            IpAddr::V6(ip) => Self::V6(ip.octets()),
+        }
+    }
+
+    const fn len(&self) -> usize {
+        match self {
+            Self::V4(_) => 4,
+            Self::V6(_) => 16,
+        }
+    }
+}
+
+impl From<AddrBytes> for [u8; 16] {
+    fn from(addr: AddrBytes) -> Self {
+        match addr {
+            AddrBytes::V4(bytes) => {
+                let mut v6 = [0; 16];
+                v6[..4].copy_from_slice(&bytes);
+                v6
+            }
+            AddrBytes::V6(bytes) => bytes,
+        }
+    }
+}
+
+#[repr(C)]
+#[derive(Default)]
+struct IfIndexMsg {
+    nlmsg: nlmsghdr,
+    rtm: rtmsg,
+    rt: rtattr,
+    addr: [u8; 16],
+}
+
+impl IfIndexMsg {
+    fn new(remote: IpAddr, nlmsg_seq: u32) -> Self {
```

Comment:
It needs to be set to a unique value for each Netlink message sent.

Reply:
Unique on what scope?  Because if it needs to be unique, you probably want a global counter rather than having it passed in to these functions.

That way, you can guarantee uniqueness rather than hard-coding numbers.  As it is, if someone calls these functions from multiple threads, I have no idea what would happen.

Reply:
Unique per socket. Happy to go with a global counter.